lib/kafka_producer.js

```diff
@@ -157,10 +244,54 @@ KafkaProducer.prototype.getWholeMsg = function getWholeMsg(topic, message, timeS
 
 KafkaProducer.prototype.close = function close(callback) {
     var self = this;
+    if (self.batching) {
+        self.flushEntireCache(callback);
+    }
     self.enable = false;
     if (self.restClient) {
         self.restClient.close();
     }
 };
 
+// The batch message payload should follow:
+// 4 BE bytes number of messages + 4 BE bytes size of message + actual message
+function MessageBatch(topic) {
```

Comment:
To be fast let's take advantage of caching.

on the `KafkaRestClient` instance you want to have a field that's an object of topic -> cacheBuffer.

A cacheBuffer can be a 100kb `new Buffer(100 * 1024)`.

Every time you allocate a `MessageBatch` we should give it this cached buffer.
When we flush a `MessageBatch` we should see how many bytes were actually written. i.e. `sizeBytes` + 4 + 4 * `numMessages` and then do `new Buffer(sizeBytes + 4 + (4 * numMessages))` and copy from the cached buffer into it.

once we've copied the data out into a new buffer we can re-use the cached buffer.

This optimization is important because concat() is expensive and it allocates new buffers every time its done.

I'd rather allocate ~10 cached buffers at startup and then allocate a single correctly-sized buffer right before flushing an entire batch.